### ISSUE TITLE
Incorrect coherence diagnostic

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#55752]
```rust
struct MyError;

impl From<i32> for Result<(), MyError> {
    fn from(x: i32) -> Self {
        if x == 0 { Ok(()) } else { Err(MyError) }
    }
}

fn main() {}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=646489805ed41f94a56e0a2077e0437d))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:3:1
  |
3 | impl From<i32> for Result<(), MyError> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
  |
  = note: the impl does not reference any types defined in this crate
  = note: define and implement a trait or new type instead

error: aborting due to previous error

For more information about this error, try `rustc --explain E0117`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

This is incorrect because the error message states:

> the impl does not reference any types defined in this crate

but the `impl` does reference one type defined in this crate: `MyError`.

**Comment by estebank** [comment]
This could be fixed by changing the wording to "the impl does not reference only types defined in this crate" either only for cases like this if easy, or always.

**Comment by lbrande** [comment]
I would like to give this a try. I haven't worked on this repo before so if possible I would appreciate some guidance.

**Comment by lbrande** [comment]
Thanks, will try to get things set up later today.

**Comment by estebank** [comment]
Another possible output (which might be easier to accomplish) could be:

```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:3:1
  |
3 | impl From<i32> for Result<(), MyError> {
  | ^^^^^----^^^^^^^^^^------^^^^^^^^^^^^^ impl doesn't use types inside crate
  |      |             |
  |      |             this type is isn't defined in this crate
  |      this trait isn't defined in this crate
  |
  = note: the impl does not reference only types defined in this crate
  = note: define and implement a trait or new type instead
```
